### PR TITLE
fix bugs and support ddp

### PR DESCRIPTION
**Pull request by Jintao-Huang**
1. add feat/ddp (todo: ddp+mp together)
2. more `.sh`
3. fix chatglm2 full infer bug (configuration.json). [issue#10](https://github.com/modelscope/swift/issues/10)
4. fix flash attention+qwen+3090 bug. [issue#9](https://github.com/modelscope/swift/issues/9)
5. fix sdk bug: is_master, Model save

### CODE REVIEW

File: examples/pytorch/llm/llm_infer.py [hunk 1]
```diff
@@ -76,18 +76,16 @@ def llm_infer(args: InferArguments) -> None:
...
+    if args.sft_type == 'full':
+        kwargs['model_dir'] = args.ckpt_dir
     model, tokenizer = get_model_tokenizer(
-        args.model_type, torch_dtype=args.torch_dtype, **model_kwargs)
+        args.model_type, torch_dtype=args.torch_dtype, **kwargs)
 
     # ### Preparing lora
     if args.sft_type == 'lora':
         model = Swift.from_pretrained(model, args.ckpt_dir)
-    elif args.sft_type == 'full':
```
Comment (tastelikefeet) [inline examples/pytorch/llm/llm_infer.py:86]: How to load checkpoint if this `full` sft_type branch is removed?

File: examples/pytorch/llm/llm_sft.py [hunk 2]
```diff
@@ -208,7 +228,17 @@ def llm_sft(args: SftArguments) -> None:
...
+        ddp_backend=args.ddp_backend,
+        gradient_checkpointing=args.gradient_checkpoint,
+        local_rank=local_rank)
+
+    if args.gradient_checkpoint:
+        # fix: gradients will be None
+        model.enable_input_require_grads()
+        if is_dist():
+            trainer_args.ddp_find_unused_parameters = False
+            trainer_args.ddp_broadcast_buffers = False
```
Comment (tastelikefeet) [inline examples/pytorch/llm/llm_sft.py:252]: broadcast_buffers set to False because?

File: examples/pytorch/llm/utils/models.py [hunk 3]
```diff
@@ -106,8 +108,17 @@ def get_model_tokenizer_qwen(model_dir: str,
...
-    setattr(model_config, k, True)
+    k_true = mapper[torch_dtype]
+    for k in mapper.values():
+        v = False
+        if k == k_true:
+            v = True
+        setattr(model_config, k, v)
+
+    device_name = torch.cuda.get_device_name()
+    if 'A100' not in device_name and 'H100' not in device_name:
```
Comment (tastelikefeet) [inline examples/pytorch/llm/utils/models.py]: How about A800? How many types of GPU support flashAttention?

File: examples/pytorch/llm/utils/utils.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+def get_dist_setting() -> Tuple[int, int, int]:
+    """return rank, local_rank, world_size"""
+    rank = int(os.getenv('RANK', -1))
+    local_rank = int(os.getenv('LOCAL_RANK', -1))
+    world_size = int(os.getenv('WORLD_SIZE', 1))
+    return rank, local_rank, world_size
+
+
+def is_dist():
+    rank = int(os.getenv('RANK', -1))
```
Comment (tastelikefeet) [inline examples/pytorch/llm/utils/utils.py:43]: When does the variable RANK exist?

File: examples/pytorch/llm/llm_infer.py (hunk 1 above)
Comment (Jintao-Huang) [reply examples/pytorch/llm/llm_infer.py:86]: We pass model_dir=ckpt_dir into get_model_tokenizer

File: swift/trainers/mixin.py [hunk 5]
```diff
@@ -3,18 +3,19 @@
...
 import shutil
 from types import MethodType
-from typing import Any, Callable, Dict, List, NewType, Optional, Tuple, Union
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
 import json
 import safetensors
 import torch
 from datasets import Dataset as HfDataset
+from modelscope import Model
```
Comment (tastelikefeet) [inline swift/trainers/mixin.py]: Don't import modelscope

File: swift/trainers/mixin.py [hunk 6]
```diff
@@ -254,6 +256,17 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
                 else:
                     torch.save(state_dict,
                                os.path.join(output_dir, 'pytorch_model.bin'))
+        elif isinstance(self.model, Model):
+            model_dir = self.model.model_dir
```
Comment (tastelikefeet) [inline swift/trainers/mixin.py]: Why not just hasattr(self.model, 'model_dir')

File: swift/utils/logger.py [hunk 7]
```diff
@@ -10,10 +10,9 @@
     '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
 
 
-def is_master(group=None):
-    from torch import distributed as dist
-    return dist.get_rank(
-        group) == 0 if dist.is_available() and dist.is_initialized() else True
+def is_master():
```
Comment (tastelikefeet) [inline swift/utils/logger.py:13]: Please move to torch_utils

Comment (Jintao-Huang) [reply swift/utils/logger.py:13]: There is the problem of recursive import need to be solved

File: swift/trainers/mixin.py (hunk 6 above)
Comment (Jintao-Huang) [reply swift/trainers/mixin.py]: if sft_type=lora, we don't need to save model type, pipeline type in configuration.json

File: swift/trainers/mixin.py (hunk 5 above)
Comment (Jintao-Huang) [reply swift/trainers/mixin.py]: sorry, that's an important question

File: examples/pytorch/llm/utils/utils.py (hunk 4 above)
Comment (Jintao-Huang) [reply examples/pytorch/llm/utils/utils.py:43]: torchrun will sets environment variables

File: examples/pytorch/llm/utils/models.py (hunk 3 above)
Comment (Jintao-Huang) [reply examples/pytorch/llm/utils/models.py]: I haven't found a better solution, perhaps passed in at the command line, but this approach breaks the overall consistency

Comment (Jintao-Huang) [reply examples/pytorch/llm/utils/models.py]: Modified to the way the command line passes in

File: swift/trainers/mixin.py (hunk 5 above)
Comment (Jintao-Huang) [reply swift/trainers/mixin.py]: fixed

**Review by tastelikefeet — APPROVED** [review]

**Pull request merged**